File: examples/extended/analysis/AnaEx01/analysis/Lab/AnaEx01.py
#
# ********************************************************************
# * DISCLAIMER                                                       *
# *                                                                  *
# * The following disclaimer summarizes all the specific disclaimers *
# * of contributors to this software. The specific disclaimers,which *
# * govern, are listed with their locations in:                      *
# *   http://cern.ch/geant4/license                                  *
# *                                                                  *
# * Neither the authors of this software system, nor their employing *
# * institutes,nor the agencies providing financial support for this *
# * work  make  any representation or  warranty, express or implied, *
# * regarding  this  software system or assume any liability for its *
# * use.                                                             *
# *                                                                  *
# * This  code  implementation is the  intellectual property  of the *
# * GEANT4 collaboration.                                            *
# * By copying,  distributing  or modifying the Program (or any work *
# * based  on  the Program)  you indicate  your  acceptance of  this *
# * statement, and all its terms.                                    *
# ********************************************************************
#
#/////////////////////////////////////////////////////////////////////////////
#/////////////////////////////////////////////////////////////////////////////
#/////////////////////////////////////////////////////////////////////////////
#
# Example of a Python script using the AnaEx01.aida file.
#
#/////////////////////////////////////////////////////////////////////////////

#
#  To be executed from the Python shell :
#   OS> <setup Lab>
#   OS> python
#  (>>> import AIDA)
#   >>> import AnaEx01
#

from AIDA import AIDA_createAnalysisFactory

aida = AIDA_createAnalysisFactory()

# Get some AIDA factories :
plotterFactory = aida.createPlotterFactory()
treeFactory = aida.createTreeFactory()
memoryTree = treeFactory.createDefault()
histogramFactory = aida.createHistogramFactory(memoryTree)

# Attach the current plotter :
plotter = plotterFactory.create('')

plotter.createRegions(1,2,0)
plotter.setParameter('title','AnaEx01 analysis')

tree = treeFactory.create('AnaEx01.aida','xml',1,0)
tree.ls()

#/////////////////////////////////////////////////////////////////////////////
# In first region, get and plot the EAbs histo :
#/////////////////////////////////////////////////////////////////////////////

tree.cd('histograms')
tree.ls()

# From an AIDA tree we retreive IManagedObjects, we have to cast.
#FIXME : IManagedObject_to_IHistogram1D is not AIDA, it is OSC specific.
from AIDA import IManagedObject_to_IHistogram1D
EAbs = IManagedObject_to_IHistogram1D(tree.find('EAbs'))
if EAbs == None:
  print 'EAbs not found or is not an AIDA::IHistogram1D.'
else :
  # Plot the histo :
  region = plotter.currentRegion()
  region.plot(EAbs)

#/////////////////////////////////////////////////////////////////////////////
# In second region plot the EAbs histo built from the tuple :
#/////////////////////////////////////////////////////////////////////////////

tree.cd('..')
tree.cd('tuples')
tree.ls()

tupleFactory = aida.createTupleFactory(tree)
# Get a tuple from the tree :
#tuple = tupleFactory.create('AnaEx01','AnaEx01','')
from AIDA import IManagedObject_to_ITuple
tuple = IManagedObject_to_ITuple(tree.find('AnaEx01'))
if tuple == None:
  print 'Tuple AnaEx01 not found or is not an AIDA::ITuple.'

# Create an histo in the default memory tree :
tuple_EAbs = histogramFactory.createHistogram1D('tuple_EAbs','AnaEx01/EAbs',100,0,100)

# Create an Evaluator and a Filter object :
evaluator = tupleFactory.createEvaluator('EAbs')
evaluator.initialize(tuple)
filter = tupleFactory.createFilter('')
filter.initialize(tuple)

# Project tuple AnaEx01/EAbs column on the tuple_EAbs histo :
tuple.project(tuple_EAbs,evaluator,filter)

plotter.next()
region = plotter.currentRegion()
region.plot(tuple_EAbs)

plotter.show()
plotter.interact()

del evaluator
del filter

print 'The two histos must give the same information !'

del tupleFactory

del tree
del plotter
del region
del plotterFactory
del treeFactory
del histogramFactory

File: examples/extended/analysis/AnaEx01/AnaEx01.cc
//
// ********************************************************************
// * DISCLAIMER                                                       *
// *                                                                  *
// * The following disclaimer summarizes all the specific disclaimers *
// * of contributors to this software. The specific disclaimers,which *
// * govern, are listed with their locations in:                      *
// *   http://cern.ch/geant4/license                                  *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.                                                             *
// *                                                                  *
// * This  code  implementation is the  intellectual property  of the *
// * GEANT4 collaboration.                                            *
// * By copying,  distributing  or modifying the Program (or any work *
// * based  on  the Program)  you indicate  your  acceptance of  this *
// * statement, and all its terms.                                    *
// ********************************************************************
//
//
// $Id: AnaEx01.cc,v 1.13 2005/12/06 11:07:17 gcosmo Exp $
// GEANT4 tag $Name: geant4-08-00 $
//
// 
// --------------------------------------------------------------
//      GEANT 4 - AnaEx01
//
// --------------------------------------------------------------
// Comments
//   Example of histogram and tuple manipulations using an AIDA compliant 
//  system. All analysis manipulations (hooking an AIDA implementation,
//  histo booking, filling, etc...) are concentrated in one 
//  class : AnaEx01AnalysisManager.
//   See the README file within the same directory to have more infos.
// --------------------------------------------------------------

// Geant4 :
#include "G4RunManager.hh"
#include "G4UImanager.hh"

#include "Randomize.hh"

// AIDA :
#ifdef G4ANALYSIS_USE
#include <AIDA/IAnalysisFactory.h>
#endif

// AnaEx01 :
#include "AnaEx01AnalysisManager.hh"
#include "AnaEx01DetectorConstruction.hh"
#include "AnaEx01PhysicsList.hh"
#include "AnaEx01PrimaryGeneratorAction.hh"
#include "AnaEx01RunAction.hh"
#include "AnaEx01EventAction.hh"
#include "AnaEx01SteppingAction.hh"
#include "AnaEx01SteppingVerbose.hh"

int main(int,char**) {

  // choose the Random engine
  CLHEP::HepRandom::setTheEngine(new CLHEP::RanecuEngine);
  
  //my Verbose output class
  G4VSteppingVerbose::SetInstance(new AnaEx01SteppingVerbose);
     
  // Construct the default run manager
  G4RunManager * runManager = new G4RunManager;

  // set mandatory initialization classes
  AnaEx01DetectorConstruction* detector = new AnaEx01DetectorConstruction;
  runManager->SetUserInitialization(detector);
  runManager->SetUserInitialization(new AnaEx01PhysicsList);

  runManager->SetUserAction(new AnaEx01PrimaryGeneratorAction(detector));

  AnaEx01AnalysisManager* analysisManager = 0;
#ifdef G4ANALYSIS_USE
  AIDA::IAnalysisFactory* aida = AIDA_createAnalysisFactory();
  analysisManager = new AnaEx01AnalysisManager(aida);
#endif
  runManager->SetUserAction(new AnaEx01RunAction(analysisManager));
  runManager->SetUserAction(new AnaEx01EventAction(analysisManager));
  runManager->SetUserAction(new AnaEx01SteppingAction(analysisManager));

  //Initialize G4 kernel
  runManager->Initialize();
    
  // get the pointer to the User Interface manager 
  G4UImanager* UI = G4UImanager::GetUIpointer();  

  // Batch mode
  UI->ApplyCommand("/control/execute run.mac");

  // job termination
#ifdef G4ANALYSIS_USE
  delete analysisManager;
#endif
  delete runManager;

#ifdef G4ANALYSIS_USE
  delete aida;
#endif

  return 0;
}

